File: 904/fruit-into-baskets.py
# 只有兩個籃子，可採2種水果，要連續採收，籃子同種水果可裝很多
# 因為陣列很長，不能用暴力法去算
# 可試試 sliding window 的作法, 不過我這裡是考 lee215 的另一種作法
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        a, b = 0, 0 # 前面是a,後面是b 
        bN = 0 # 後面的b連續的個數
        ans, cur = 0, 0 # 最後答案，目前答案
        # for i in range(1,N):
        for now in fruits: # 逐項拿出水果來檢查
            if now==b: # 和右邊項相同，b繼續連續，就是變長的意思
                cur += 1 # 目前答案變長
                bN += 1 # b 的數目變多，連續的b有幾個
            elif now==a: # 和左邊項相同，左右要交換
                cur += 1 # 答案還是會變長
                a, b = b, a # 左右要交換
                bN = 1 # b重新從1開始，連續的b有幾個
            else: # now 不是 a 也不是 b, 要把 a 淘汰掉
                cur = bN + 1 # 把 連續的b的長度，加now這1個
                a, b = b, now # 要更新 a,b 項
                bN = 1 # b重新從1開始，連續的b有幾個
            ans = max(ans, cur)
        return ans

File: 904/fruit-into-baskets.c
// 你只有2個籃子，即「只能採集」2種水果。
// 從 fruits[i] 開始往右採集，只能採集2種水果，可採集多少水果。
// 可使用 sliding window 來處理。我模仿之前寫的 Python 程式, 改用 C 實作
int totalFruit(int* fruits, int fruitsSize) {
    int a = 0, b = 0, bN = 0; //a是前一籃的水果編號, b是後一籃的水果編號, bN是b水果連續累積幾個
    int ans = 0, cur = 0; //ans最後的答案, cur目前這個的答案
    for(int i=0; i<fruitsSize; i++){
        if(fruits[i]==b){ //與最後一籃的b水果相同
            cur ++; //現在的答案變長
            bN ++; // b水果+1
        }else if(fruits[i]==a){ //與前一籃的a水果相同
            cur ++; //現在的答案變長
            int temp = a;
            a = b;
            b = temp; //將兩籃交換
            bN = 1; //重新從1開始, 目前 b 連續 1個
        }else{ //不幸fruits[i]與a,b都不相同, 要捨棄a了
            cur = bN + 1; //之前的b變a,長度+現在新水果1個
            a = b; //b變成前一籃的a
            b = fruits[i]; //新的b
            bN = 1; ////重新從1開始, 目前 b 連續 1個
        }
        if(cur>ans) ans = cur;
    }
    return ans;
}
